### DISCUSSION TITLE
Welcome to traffic Discussions!

### DISCUSSION THREAD

**Discussion by xoolive** [xoolive/traffic#112] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by luis-uow** [comment]
Hi @xoolive,

Many thanks for developing this repository. It looks like a very nice interface to use python with OpenSky. This is the first time we are using OpenSky for one of our research projects, so we are still assessing which tools to use, and how to access and process the ADS-B data.

A couple of quick questions I hope you can help us with:

- When downloading the data from OpenSky, I see how you can do polygons to clip the trajectories and only download the ones within those boundaries, but can you also modify the time-step? It is to try to reduce the downloading size and time.

- What would be the best way to add capabilities to the processing of flight functions you have already developed? For example, what would be the best way to add a filter to select the segment of the trajectories which are lower than a given FL at arrival? I see in the code how you have iterators to identify segments which correspond to holdings, for example. Would the best way to do something similar within the NavigationFeatures class?

Many thanks!

P.S. Also, not sure if this discussion page is the best for these comments...

**Reply by xoolive** [reply]
Hi @luis-uow,

Thank you for opening the discussions (indeed since there are no issue, I believe here is best 😃 )

## Downloading from Opensky Impala shell

- You got a few helper functions [here](https://traffic-viz.github.io/opensky_impala.html). The `history` is usually the most used one. All the functions generate the proper SQL query, splitting in chunks of hours (or more, see `date_delta`), based on the `request` function ▶️  [documentation](https://traffic-viz.github.io/opensky_impala.html#custom-request)

- You will find in the [code file](https://github.com/xoolive/traffic/blob/master/traffic/data/adsb/opensky_impala.py) how things are made for the basic functions

- You can basically write any query with the `request` function, esp. the particular recipes on the [official Impala page](https://opensky-network.org/impala-guide) For your particular need, try Ctrl+F "quotient" on that page

- If you think it could be a new function on its own, or a new argument in the `history` function, PR are more than welcome!

## New capabilities

I guess it depends... 🤦 

- *select the segment of the trajectories which are lower than a given FL at arrival*  
  For these ones, I usually don't add new capabilities: I would just write something like the following function to select only trajectories flying below a given altitude for the last five minutes of available data, then [stack the operation](https://traffic-viz.github.io/quickstart.html?highlight=max_workers#lazy-iteration) on batches of data.
```python
def lower_arrival(flight, flight_level: float) -> bool:
    return flight.last("5T").altitude_max < 100 * flight_level
```

- Indeed, I am working on a decent holding selection method, "smart", so this one would probably be in the `NavigationFeatures` (just a mixin to add more functions to `Flight`)

- If you implement very particular methods which are a pain in the a** to get right (like *detecting holding positions on a taxiway* for example), this would go there as well.

Please, also, feel free to question any choice of implementation in those methods. That's always the best I could come up with (I am not withholding anything), but I don't pretend it's the Holy Gospel. If there are many ways to get something right, we can add an option to select the method. If one option is better than the other, we can remove the older one. The only bottom line is to agree on the semantics behind each method.

**Reply by xoolive** [reply]
Also, we can arrange some discussion time if you prefer to talk about it. You know where to find me...

**Reply by luis-uow** [reply]
Hi @xoolive,

Thanks for your quick and useful replies. I'll try your suggestions on how to query the data and filter the trajectories. I might take your offer and we can maybe have a chat in the future (once I explore the library a bit more so I don't ask obvious questions :-) )

**Comment by ElSabio97** [comment]
Hello @xoolive !

I am currently doing a dissertation for my degree and is about RA events over EUROCONTROL airspace. I would like to know if it would be possible to add something similar to a progress bar to know how far in the download process one is. I believe this would be very useful in the case of a huge download. It would be something similar to the output this @junzis code gives but with a progress bar:

```python
from pyopensky import OpenskyImpalaWrapper

opensky = OpenskyImpalaWrapper()
# Perform a simple and massive query (~20k records for 1 second here!)
df = opensky.query(type="raw", start="2018-07-01 13:00:00", end="2018-07-01 13:00:01")
```
For "output" I mean the text you get on the console, not variables or datafiles.

Finally, this is an appreciation message for your work:

Thank you very much for what you have created in this repo. It is truly amazing, useful and intuitive. I can sincerely say it because one month ago I had never ever heard of Spyder nor Anaconda or even written a single line of Python, and now I'm still very far from edditing your repo to customize it for my work but I enjoy processing data and appreciate your work and other's in the OpenSky Network.

Best!

**Reply by xoolive** [reply]
Hello @ElSabio97 
I am happy to hear you are finding your way with the library, thank you for your kind words.
The progress bar you mention is supposed to be already integrated if the query you send is processed piecewise. Do you have any particular query where you don't see a progress bar?

**Reply by ElSabio97** [reply]
Oh thats weird. Maybe it is because I'm using Spyder?

I am working with this:
```python
from traffic.data import opensky

df = opensky.rawdata(
    start="2021-02-01 18:00",
    stop="2021-02-01 18:05",
    callsign = "IB%"    
)
```
And the output in the console is empty.

Also, reading what you said to @luis-uow about the optional argument `date_delta`, I would like to know how the proper coding would be to minimize the download time on the example I wrote above. Should I request a higher `timedelta` or lower? 

For instance I am trying this now: 

```python
df = opensky.rawdata(
    start="2021-02-01 17:00",
    stop="2021-02-01 17:05",
    date_delta = timedelta(minutes=1),
    callsign = "IB%"    
)
```
But it takes longer than the default option _per hour_.

**Reply by xoolive** [reply]
Oh it's normal: the query is not split here (by default, split per hour, see the `date_delta` parameter). And there is no way to know how far you are in the downloading during one query...

If you want a richer (noisy) output, try setting a different logging level:
```python
import logging
logging.basicConfig(level=logging.INFO)  # or logging.DEBUG
```

**Reply by xoolive** [reply]
The splitting is a trade-off indeed. More queries means more blocking time, but less queries means more volume to process when you get the data (and you can be limited by your RAM). Here I would stay with the default option *per hour*.

**Comment by fraciocca** [comment]
Hi @xoolive!
I'm Francesco and I'm an Aeronautical student at Politecnico of Milan. I will graduate in the next July and I'm working on my thesis, which, in a very few words, is about data fusion between different position data sources that are gathering ADS-B and FLARM messages from any other air vehicle. 
Firstly I want to thank you very much for sharing this amazing tool that makes the life easier for students like me.
I just wanted to ask you a simple question about this project and if what I'm searching for is feasible or not.
I've applied for the OpenSky's Impala shell historical dataset and I've been granted access.
I would like to know if it's possible, through your tool, to download ADS-B messages from any air vehicle inside a certain area that I would fix (for example in the proximity of a certain Airport).
I've read the documentation you have provided (by the way, it's very useful and self-explanatory!) and I would like if you could explain me how could I achieve what I'm looking for, how this data can be saved and in which format (because I would need to feed this messages inside a Matlab script).
I hope I did not miss anything from the documentation, but I also think that asking my doubts right here could also help, in the future, people that have some similar issues.
Best

**Reply by xoolive** [reply]
Thank you for your message.
Please check the parameters for the `opensky.history` function [here](https://traffic-viz.github.io/data_sources/opensky_impala.html#traffic.data.adsb.opensky_impala.Impala.history)
You can also read about formats [here](https://traffic-viz.github.io/data_sources/export.html)

**Comment by JRO018** [comment]
Hi @xoolive!

I'm Josep and I'm an Aerospace student at Escola d'Enginyeria de Telecomunicació i Aeroespacial de Castelldefels. I will finish my studies in July (hopefully) and I'm working on my thesis which is about data recollection from LEBL and study of the behaviour of different kind of operations and configurations. By the recommendation of my thesis supervisor I'm going to use traffic for the data recollection and study of the trajectories.

I'm kinda new into conda environments and I've never used a library like this before. I've "succesfully" installed the traffic library in the Anaconda.navigator  and I've been able to call sample trajectories but when it comes to calling data from opensky I haven't been able to obtain nothing in return.

I have executed the following code:

```python
from traffic.data import eurofirs
from traffic.data import opensky
t_aib = opensky.history(
    start="2019-11-01 09:00",
    stop="2019-11-01 12:00",
    departure_airport="LEBL",
    arrival_airport="LEBL",
)
```
And the response have been this one:
```
--------------------------------------------------------------------------
SSHException                              Traceback (most recent call last)
...
SSHException: No authentication methods available
```

I know this is may be a specific case and maybe is not the intended purpose of this conversation but I need help on how to solve this "problem" and there is no better advice than the one given by the creator itself. Thank you in advance!

**Reply by xoolive** [reply]
Hi @JRO018 

I have not done a release yet with the final updates (wrapping [`pyopensky`](https://github.com/open-aviation/pyopensky)) but few things:
- have you been granted access to the Trino database? [here](https://open-aviation.github.io/pyopensky/credentials.html)
- your code will download trajectories *to* and *from* `LEBL` (not *or*), you might want to switch to `airport="LEBL"`
- an `SSHException` suggests that you are trying to connect the Impala shell, which is deprecated now. Check your [configuration files](https://open-aviation.github.io/pyopensky/credentials.html#configuration)

**Reply by JRO018** [reply]
Hi @xoolive 

In first place thank you for your fast answer, I've followed the steps you specified in your post and I've granted myself access to the Trino database today, I have also modified the piece of code according to your specifications but now when running the code this error comes up.

---------------------------------------------------------------------------
NoValidConnectionsError                   Traceback (most recent call last)
Cell In[4], line 3
      1 from traffic.data import eurofirs
      2 from traffic.data import opensky
----> 3 t_aib = opensky.history(
      4     start="2019-11-01 09:00",
      5     stop="2019-11-01 12:00",
      6     airport="LEBL",
      7 )

File /opt/anaconda3/envs/traffic/lib/python3.10/site-packages/traffic/data/adsb/opensky.py:384, in OpenSky.history(self, start, stop, callsign, icao24, serials, bounds, departure_airport, arrival_airport, airport, cached, compress, limit, return_flight, *args, **kwargs)
    361 @copy_documentation(trino.Trino.history)
    362 def history(
    363     self,
   (...)
    382     **kwargs: Any,
    383 ) -> None | Flight | Traffic:
--> 384     df = self.db_client.history(
    385         start,
    386         stop,
    387         *args,
    388         callsign=callsign,
    389         icao24=icao24,
    390         serials=serials,
    391         bounds=bounds,
    392         departure_airport=departure_airport,
    393         arrival_airport=arrival_airport,
    394         airport=airport,
    395         cached=cached,
    396         compress=compress,
    397         limit=limit,
    398         **kwargs,
    399     )
    400     if df is None:
    401         return None

File /opt/anaconda3/envs/traffic/lib/python3.10/site-packages/pyopensky/impala.py:976, in Impala.history(self, start, stop, callsign, icao24, serials, bounds, departure_airport, arrival_airport, airport, time_buffer, cached, compress, limit, other_tables, other_params, progressbar, date_delta, count, *args, **kwargs)
    953 _log.info(
    954     f"Sending request between time {bt} and {at} "
    955     f"and hour {bh} and {ah}"
    956 )
    958 request = self.basic_request.format(
    959     columns=columns,
    960     before_time=bt.timestamp(),
   (...)
    973     where_clause=where_clause,
    974 )
--> 976 df = self._impala(
    977     request, columns=parse_columns, cached=cached, compress=compress
    978 )
    980 if df is None:
    981     continue

File /opt/anaconda3/envs/traffic/lib/python3.10/site-packages/pyopensky/impala.py:335, in Impala._impala(self, request, columns, cached, compress)
    333 if not self.connected:
    334     _log.info("Connecting the database")
--> 335     self._connect()
    337 # bug fix for when we write a request with """ starting with \n
    338 request = request.replace("\n", " ")

File /opt/anaconda3/envs/traffic/lib/python3.10/site-packages/pyopensky/impala.py:298, in Impala._connect(self)
    295     _log.info(f"Using ProxyCommand: {self.proxy_command}")
    296     extra_args["sock"] = paramiko.ProxyCommand(self.proxy_command)
--> 298 client.connect(
    299     "data.opensky-network.org",
    300     port=2230,
    301     username=self.username,
    302     password=self.password,
    303     look_for_keys=False,
    304     allow_agent=False,
    305     compress=True,
    306     **extra_args,  # type: ignore
    307 )
    308 self.stdin, self.stdout, self.stderr = client.exec_command(
    309     "-B", bufsize=-1, get_pty=True
    310 )
    311 self.connected = True

File /opt/anaconda3/envs/traffic/lib/python3.10/site-packages/paramiko/client.py:409, in SSHClient.connect(self, hostname, port, username, password, pkey, key_filename, timeout, allow_agent, look_for_keys, compress, sock, gss_auth, gss_kex, gss_deleg_creds, gss_host, banner_timeout, auth_timeout, channel_timeout, gss_trust_dns, passphrase, disabled_algorithms, transport_factory, auth_strategy)
    403     # Make sure we explode usefully if no address family attempts
    404     # succeeded. We've no way of knowing which error is the "right"
    405     # one, so we construct a hybrid exception containing all the real
    406     # ones, of a subclass that client code should still be watching for
    407     # (socket.error)
    408     if len(errors) == len(to_try):
--> 409         raise NoValidConnectionsError(errors)
    411 if transport_factory is None:
    412     transport_factory = Transport

NoValidConnectionsError: [Errno None] Unable to connect to port 2230 on 194.209.200.3

This ip is neither the ip of my machine nor my public ip so I don't understand where this value is coming from, again thank you in advance for your patience and predisposition :)

**Reply by xoolive** [reply]
Could you do a pip update to the latest release of traffic and try again? (Conda doesn't work yet...)

**Reply by JRO018** [reply]
I've done and update to the latest release using "conda update -n traffic -c conda-forge traffic" and also using pip, none of them have achieved a successful result and with the pip option this error appears :(

![Captura de pantalla 2024-04-16 a las 20 31 20](https://github.com/xoolive/traffic/assets/167023740/f36a927b-e557-4d37-8863-607c3e73af03)

**Reply by xoolive** [reply]
Can you confirm your Python version is at least 3.9 ? (I recommend you pick 3.11 or 3.12 to be honest)

**Reply by JRO018** [reply]
I've done a reinstall and everything is working fine, thank you for your help!

**Reply by JRO018** [reply]
Regarding some functions included in traffic, I've been experimenting with "aligned_on_ils" and trying some of the examples displayed on the traffic homepage. When it comes to the use of this function do I need any previous filtering on the data extracted from opensky historical data or I can use directly? 

Just to put more context into the situation, I've done the download of data like this into a csv file:

from traffic.data import eurofirs
from traffic.data import opensky
import csv

elprat = opensky.history(
    start="2023-01-01 12:00",
    stop="2023-01-01 15:00",
    airport="LEBL",
)

elprat.to_csv("prueba_15.csv")

And in order to do some testing I've run a code like this to study the header assigned to some flights:

import traffic
from traffic.core import Flight
import traffic.data
datos = Flight.from_file("prueba_15.csv")

#Hacemos perímetro exterior del aeropuerto para estudiar solo los vuelos concevido dentro de la caja perimetral
max_altitude = 1000 # Dterminamos la altitud máxima del perímetro del aeropuerto

min_lat, max_lat = 41.26133889 , 41.33023611 # Latitude mínima y máxima de la caja
min_lon, max_lon = 2.03469722, 2.14402778  # Longitude mínima y máxima de la caja

#Visualizamos cuántos vuelos tenemos dentro del perímetro

filtered_flights = datos.query(f"{min_lat} <= latitude <= {max_lat} and {min_lon} <= longitude <= {max_lon} and geoaltitude <= {max_altitude}")
print("Vuelos dentro de la caja")
print(filtered_flights)

#Creamos un perímetro para la cabecera 06L General

min_lat, max_lat = 41.29171944 , 41.29843611  # Example latitude boundaries 06L
min_lon, max_lon = 2.06516667, 2.08072222  # Example longitude boundaries 06L
max_altitude = 300# Example maximum altitude

# Filter flights based on latitude, longitude, and altitude
filtered_flights_06L = filtered_flights.query(f"{min_lat} <= latitude <= {max_lat} and {min_lon} <= longitude <= {max_lon} and geoaltitude <= {max_altitude}")
print("Vuelos 06L")
print(filtered_flights_06L)


#Creamos un perímetro para la cabecera 24R Despegues

min_lat, max_lat = 41.30270833 , 41.30663056  # Example latitude boundaries 24R
min_lon, max_lon = 2.09662500, 2.10424167  # Example longitude boundaries 24R
max_altitude = 0 # Example maximum altitude

# Filter flights based on latitude, longitude, and altitude
filtered_flights_24R = filtered_flights.query(f"{min_lat} <= latitude <= {max_lat} and {min_lon} <= longitude <= {max_lon} and geoaltitude <= {max_altitude}")
print("Vuelos 24R")
print(filtered_flights_24R)

#Creamos un perímetro para la cabecera 24R Aterrizajes

min_lat, max_lat = 41.30270833 , 41.30663056  # Example latitude boundaries 24R
min_lon, max_lon = 2.09662500, 2.10424167  # Example longitude boundaries 24R
max_altitude = 1000 # Example maximum altitude

# Filter flights based on latitude, longitude, and altitude
filtered_flights_24R = filtered_flights.query(f"{min_lat} <= latitude <= {max_lat} and {min_lon} <= longitude <= {max_lon} and geoaltitude <= {max_altitude}")
print("Vuelos 24R")
print(filtered_flights_24R)

In general I have experienced through the execution of the code and visualizations of the csv file that the results are not perfect since some flights have unlogical altitude and geoaltitudes assigned to their position and this gives errors to the outcome of the filtering. Since my goal is to filter the flights depending on the header and type of operation this is making me wonder if I would be able to use aligned_on_ils and takeoff_from_runway functions in order to obtain directly the runway header, the type of operation and the time the operation was executed in order to do some later calculations based on this data.

**Reply by xoolive** [reply]
Your message is too long, I don't know where the question is. I also see some interrogations that get solved if you just read through the documentation first
